pkg/volume/glusterfs/glusterfs.go

```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	return nil
 }
 
+func (r *glusterfsVolumeProvisioner) ConstructBitmap() error {
+
+	pvList, err := r.plugin.host.GetKubeClient().Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		glog.Errorf("glusterfs: failed to get existing persistent volumes")
+		return fmt.Errorf("glusterfs: failed to get existing persistent volumes")
+	}
+	// Collect the SC names and allocated GIDs from existing PVs.
+	scNameGids := make(map[string][]string)
+	for _, pv := range pvList.Items {
+		if pv.Annotations[annDynamicallyProvisioned] == glusterfsPluginName {
+			scNameInAnnot := pv.Annotations[annStorageClass]
+			scNameGids[scNameInAnnot] = append(scNameGids[scNameInAnnot], pv.Annotations[volumehelper.VolumeGidAnnotationKey])
+		}
+	}
+	glog.V(3).Infof("glusterfs: PVs with SC name and allocated GID Map :%v ", scNameGids)
+
+	for key := range scNameGids {
+		scdetails, _ := r.plugin.host.GetKubeClient().Storage().StorageClasses().Get(key)
+		if scdetails.Parameters["gidMin"] == "" {
+			scdetails.Parameters["gidMin"] = defaultGidMin
+		}
+		if scdetails.Parameters["gidMax"] == "" {
+			scdetails.Parameters["gidMax"] = defaultGidMax
+		}
+		blockStartnew32, err := r.ConvertGid(scdetails.Parameters["gidMin"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMin"])
+		}
+		blockEndnew32, err := r.ConvertGid(scdetails.Parameters["gidMax"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMax"])
+		}
+		// For each SC, create a pool and store it in plugin allocator map.
+		if r.plugin.gidAllocators[key] == nil {
+			r.plugin.gidAllocators[key], err = gidPoolCreate(blockStartnew32, blockEndnew32)
+			if err != nil {
+				glog.Errorf("glusterfs: gid pool creation failed %v", err)
+				return fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+		}
+	}
+
+	// For each SCs in the allocator, set bitmap for existing/previosly allocated GIDs.
+	for name, gids := range scNameGids {
+		for _, gid := range gids {
+			tempGid, err := r.ConvertGid(gid)
+			if err == nil {
+				if r.plugin.gidAllocators[name] != nil {
+					err = r.plugin.gidAllocators[name].Allocate(Block{Start: tempGid, End: tempGid})
+				}
+			} else {
+				glog.Errorf("glusterfs: failed to convert gid %s", gid)
+			}
+		}
+	}
+	return nil
+}
+
 func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 	var err error
+	// Controller restart is handled here with initialized bool in plugin struct.
+	if !r.plugin.initialized {
+		err = r.ConstructBitmap()
```

Comment:
@liggitt @obnoxxx  To summarize, at present  `ConStructAllocators` ( renamed from `ConstructBitmap`) is invoked only when there is a controller restart. In that func, we mainly do below.

(LOCK) -  ( Not available in patch yet)
1) Filter gluster PVs and fetch all the allocated GIDs in each storage class.
2) create a pool for above storage classes and record these GIDs in each of them.
(UNLOCK) -  ( Not available in patch yet)

If we `LOCK and UNLOCK` as shown above, I think the only concern here is about the Storage classes which got `deleted and recreated` (obv. with same class name) with a different `gidMin and gidMax` after we recreated the pool. As the allocator's pointer is  stored in plugin struct and whenever there is a new provision request comes in, it *only* validate the allocator existence and ignore the new gidMin/gidMax ( pool range) carried in this request. However when `AllocateNext()` is called from this new provision instance it should get an error if that GID does not belong to the pool range of existing allocator. In short, the new values of pool range is not respected if there was a `storage class recreation` and `not edit`.